File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-06/estruturais/Bridge/bridge01.py
from abc import ABC, abstractmethod

# IMPLEMENTAÇÃO DAS CORES
# interface das cores
class Cor(ABC):
    @abstractmethod
    def getCor(self):
        pass

# implementação das cores:
class VermelhoCor(Cor):
    def getCor(self):
        return "vermelho"
    
class AzulCor(Cor):
    def getCor(self):
        return "azul"
    
class VerdeCor(Cor):
    def getCor(self):
        return "verde"
    
# IMPLEMENTAÇÃO DAS FORMAS
class Forma(ABC):
    def __init__(self, cor):
        self.cor = cor

    @abstractmethod
    def desenho(self):
        pass

# Abstração refinada:
class Circulo(Forma):
    def desenho(self):
        print(f"desenhando um circulo {self.cor.getCor()}")

class Quadrado(Forma):
    def desenho(self):
        print(f"desenhando um quadrado {self.cor.getCor()}")

class Triangulo(Forma):
    def desenho(self):
        print(f"desenhando um triangulo {self.cor.getCor()}")

# cliente
vermelho = VermelhoCor()
azul = AzulCor()
verde = VerdeCor()

estrutura1 = Circulo(vermelho)
estrutura2 = Circulo(verde)
estrutura3 = Circulo(azul)
estrutura4 = Circulo(vermelho)
estrutura5 = Quadrado(verde)
estrutura6 = Triangulo(azul)

estrutura1.desenho()
estrutura2.desenho()
estrutura3.desenho()
estrutura4.desenho()
estrutura5.desenho()
estrutura6.desenho()

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-06/estruturais/Bridge/bridge01.js
// IMPLEMENTAÇÃO DAS CORES
// interface das cores:
class Cor {
    constructor(cor) {
        this.cor = cor;
    }

    getCor() {
        return this.cor;
    }
}

// concreta das cores:
class VermelhoCor extends Cor {
    constructor() {
        super("vermelho");
    }
}

class AzulCor extends Cor {
    constructor() {
        super("azul");
    }
}

class VerdeCor extends Cor {
    constructor() {
        super("verde");
    }
}

// IMPLEMENTAÇÃO DAS FORMAS\:
// interface das formas:
class Forma {
    constructor(cor) {
        this.cor = cor;
    }

    setCor(cor) {
        this.cor = cor;
    }

    desenho() {
        throw new Error("este metodo precisa ser implementado pela subclasse")
    }
}

// Abstrações refinadas:
class Circulo extends Forma {
    desenho() {
        console.log(`desenhando um circulo ${this.cor.getCor()}.`)
    }
}
class Quadrado extends Forma {
    desenho() {
        console.log(`desenhando um quadrado ${this.cor.getCor()}.`)
    }
}
class Triangulo extends Forma {
    desenho() {
        console.log(`desenhando um triangulo ${this.cor.getCor()}.`)
    }
}

// cliente: 
const vermelho = new VermelhoCor();
const azul = new AzulCor();
const verde = new VerdeCor();

const estrutura1 = new Circulo(vermelho);
const estrutura2 = new Circulo(verde);
const estrutura3 = new Circulo(azul);
const estrutura4 = new Circulo(vermelho);
const estrutura5 = new Quadrado(verde);
const estrutura6 = new Triangulo(azul);

estrutura1.desenho();
estrutura2.desenho();
estrutura3.desenho();
estrutura4.desenho();
estrutura5.desenho();
estrutura6.desenho();